timemachine/exchange/exchange_mover.py

```diff
@@ -0,0 +1,354 @@
+    return Forcefield(
+        ff.hb_handle,
+        ff.ha_handle,
+        ff.pt_handle,
+        ff.it_handle,
+        q_handle=q_handle,
+        q_handle_solv=q_handle_solv,
+        q_handle_intra=q_handle_intra,
+        lj_handle=ff.lj_handle,
+        lj_handle_solv=ff.lj_handle_solv,
+        lj_handle_intra=ff.lj_handle_intra,
+        protein_ff=ff.protein_ff,
+        water_ff=ff.water_ff,
+    )
+
+
+def test_exchange():
+    parser = argparse.ArgumentParser(description="Test the exchange protocol in a box of water.")
+    parser.add_argument("--water_pdb", type=str, help="Location of the water PDB", required=True)
+    parser.add_argument("--ligand_sdf", type=str, help="SDF file containing the ligand of interest", required=True)
+
+    args = parser.parse_args()
+
+    suppl = list(Chem.SDMolSupplier(args.ligand_sdf, removeHs=False))
+    mol = suppl[0]
+
+    ff = setup_forcefield()
+
+    seed = 2024
+
+    np.random.seed(seed)
+
+    nb_cutoff = 1.2  # this has to be 1.2 since the builders hard code this in (should fix later)
+    initial_state, nwm, topology = get_initial_state(args.water_pdb, mol, ff, seed, nb_cutoff)
+    ctxt = initialize_ctxt(initial_state)
+
+    # set up water indices, assumes that waters are placed at the front of the coordinates.
+    water_idxs = []
+    for wai in range(nwm):
+        water_idxs.append([wai * 3 + 0, wai * 3 + 1, wai * 3 + 2])
+
+    water_idxs = np.array(water_idxs)
+    bps = initial_state.potentials
+
+    # [0] nb_all_pairs, [1] nb_ligand_water, [2] nb_ligand_protein
+    nb_beta = bps[-1].potential.potentials[1].beta
+    nb_cutoff = bps[-1].potential.potentials[1].cutoff
+    nb_ligand_water_params = bps[-1].potential.params_init[1]
+
+    print("number of water atoms", nwm * 3, "number of ligand atoms", mol.GetNumAtoms())
+    print("ligand_water parameters", nb_ligand_water_params)
+
+    mover = ExchangeMover(nb_ligand_water_params, nb_beta, nb_cutoff, water_idxs)
+    cur_box = initial_state.box0
+    cur_x_t = initial_state.x0
+    cur_v_t = np.zeros_like(cur_x_t)
+
+    seed = 2023
+    writer = cif_writer.CIFWriter([topology, mol], "water.cif")
+    cur_x_t = image_frames(initial_state, [cur_x_t], [cur_box])[0]
+    writer.write_frame(cur_x_t * 10)
+
+    # equilibration
+    print("Equilibrating the system...", end="", flush=True)
+    cur_x_t, cur_v_t, cur_box = propagate(ctxt, cur_x_t, cur_v_t, cur_box, 50000)
+    print("Done")
+
+    # TBD: cache the minimized and equilibrated initial structure later on to iterate faster.
+
+    cur_x_t = image_frames(initial_state, [cur_x_t], [cur_box])[0]
+    writer.write_frame(cur_x_t * 10)
+
+    md_steps_per_batch = 1000
+    for idx in range(100000):
+        density = compute_density(nwm, cur_box)
+        cur_x_t = image_frames(initial_state, [cur_x_t], [cur_box])[0]
+        occ = compute_occupancy(cur_x_t, cur_box, initial_state.ligand_idxs, threshold=0.46)
```

Comment:
(note to self: will just need to adapt this `threshold` if using a different spherical cage (C60 fullerene vs. C90 fullerene vs. ...), or modify signature (e.g. to accept `cage_idxs` and compute `threshold = median(distance(centroid, x[cage_idxs], box))` or similar).)

Reply:
Yes - this was computed using a heuristic. A C90 is an ellipsoid as it turns out, so this was computed by measuring the diameter along the widest dimension of the ellipsoid, and halving it into the radius.